### PR TITLE
Allows custom tensors for target

### PR DESCRIPTION
**Pull request by Dref360**
This would help a lot of people who are working with Tensorflow's IO. 

For example, here is a solution to use FIFOQueue with Keras.
https://gist.github.com/Dref360/2c8280fc497df690ebb35646f7021a62

### CODE REVIEW

**Comment by fchollet** [comment]
At this point, wouldn't it be much more reasonable not to rely on `fit`/etc?

**Comment by Dref360** [comment]
Imo it's a small addition to add a lot of flexibility to Keras. Otherwise, people would have to use a lower level API and/or tweak Keras like I did. The thing I like with compile and fit it's that you can use calback you do not worry about the Optimizer, etc. 

Your call I guess :P

**Comment by Dref360** [comment]
Also if the shape of y_true is different from y_pred, you can feed the actual tensors. (Like in the YOLO's loss where y_pred shape is different)

**Comment by ahundt** [comment]
Could this work with data loaded from a TFRecord if an appropriate target is supplied? This looks very useful!

If it is merged could the fifomodel + example be added to an appropriate location either here or in [keras-contrib](https://github.com/farizrahman4u/keras-contrib)?

**Comment by Dref360** [comment]
TFRecord works pretty much the same way as queue so there is no reason why this shouldn't work. 
Yeah I'll post some example in keras-contrib if this is accepted. 

@fchollet final decision please. The workaround without this PR is to not use compile and building yourself the train_loop

**Comment by ahundt** [comment]
I think it might also be worth improving the name + docstring of the `targets` parameter, the purpose and usage is a little unclear. Would this also be compatible with other backends conceptually?

There seems to be some similar precedent with the [input_tensor param](https://github.com/fchollet/keras/search?p=2&q=%22input_tensor%3D%22&type=&utf8=%E2%9C%93), though this is mostly found in examples/applications.

**Comment by adammenges** [comment]
This would actually be pretty nice 👍

**Comment by fchollet** [comment]
After some time to think about it, I believe this is a good feature to have. However, `targets` would need to support single tensors, tensor lists, and tensor dictionaries, much like `fit`/etc.

If you are interested, you can reopen a PR, otherwise I will handle it. Please let me know.

**Comment by ahundt** [comment]
Oh now that I see this again, it is just like my original input tensor api proposal. This is very similar to what one of my old versions of #6928 does, it would be easy to re-incorporate to the current version and I think it accounts for a few more situations. 

Honestly though if it works I'm happy to have anything that will do the trick integrated into master, no matter who writes it. :-)

**Comment by ahundt** [comment]
Here is the old version which @Dref360 had approved, though labels would be sensibly renamed targets and some other key naming and design improvements have been done since then: https://github.com/fchollet/keras/pull/6928/files/9dd1ea9b8fc6189ba9475225f17ba4dbb74df95e

**Comment by fchollet** [comment]
I think we should use the name "target_tensors" for maximum explicitness,
and allow single tensors/a list/a dict.

This feature will be useful to implement "step 2" in the data tensor action
plan.

**Comment by ahundt** [comment]
@fchollet implemented in 353bb12 which is part of #6928, with support for single tensors/a list/ a dict, plus extra unit tests in addition to what was in this PR.
